evaluation.py

```diff
@@ -97,7 +98,40 @@ def save_stacked_evaluation_heatmaps(self, h, X, y, stacked_predict_heatmaps_fil
         img_v_resize = self._vstack_images(heatmap_imgs) 
         
         cv2.imwrite(filename, img_v_resize) 
-    
+
+    # Resources for heatmaps to keypoints
+    # https://github.com/yuanyuanli85/Stacked_Hourglass_Network_Keras/blob/eddf0ae15715a88d7859847cfff5f5092b260ae1/src/eval/heatmap_process.py#L5
+    # https://github.com/david8862/tf-keras-stacked-hourglass-keypoint-detection/blob/56707252501c73b2bf2aac8fff3e22760fd47dca/hourglass/postprocess.py#L17
+   
+    ### Returns np array of predicted keypoints from one image's heatmaps
+    def heatmaps_to_keypoints(self, heatmaps, threshold=HM_TO_KP_THRESHOLD):
+        keypoints = list()
+        for i in range(NUM_COCO_KEYPOINTS):
+            hmap = heatmaps[:,:,i]
+            # Resize heatmap from Output DIM to Input DIM
+            resized_hmap = cv2.resize(hmap, INPUT_DIM, interpolation = cv2.INTER_AREA)
+            # Do a heatmap blur with gaussian_filter
+            resized_hmap = gaussian_filter(resized_hmap, HEATMAP_SIGMA)
+
+            # Get peak point (brightest area) in heatmap with 3x3 max filter
+            peaks = self._non_max_supression(resized_hmap, threshold, windowSize=3)
+
+            # Choose the max point in heatmap (we only pick 1 keypoint in each heatmap)
+            # and get its coordinates and confidence
+            y, x = np.where(peaks == peaks.max())
+            if len(x) > 0 and len(y) > 0:
+                keypoints.append((int(x[0]), int(y[0]), peaks[y[0], x[0]]))
+            else:
+                keypoints.append((0, 0, 0))
```

Comment:
I may be mistaken, but won't `np.where(peaks == peaks.max())` always return and x, y (since every array has a max value)? Has this been tested to mark a weak heatmap as an unlabelled kp?

Reply:
The unlabelled kp's are turned to 0 under the threshold in `_non_max_supression` which is why we may have to test out the threshold as we go. It will label 0,0,0 for any heatmap under the given threshold.